File: core/internal/serialization/serialization.go
package serialization

import (
	"encoding/binary"
	"encoding/json"
	"errors"
	"fullstackedorg/fullstacked/types"
	"math"
	"strconv"
)

func NumberToUin4Bytes(num int) ([]byte, error) {
	if num < 0 {
		return nil, errors.New("cannot convert negative number to uint 4 bytes")
	}

	if num > types.MAX_UINT_4_BYTES {
		return nil, errors.New("converting too high number to uint 4 bytes")
	}

	bytes := make([]byte, 4)
	bytes[0] = uint8((uint(num) & uint(0xff000000)) >> 24)
	bytes[1] = uint8((num & 0x00ff0000) >> 16)
	bytes[2] = uint8((num & 0x0000ff00) >> 8)
	bytes[3] = uint8((num & 0x000000ff) >> 0)
	return bytes, nil
}
func Uint4BytesToNumber(bytes []byte) (int, error) {
	if bytes == nil {
		return 0, errors.New("cant convert nil to int")
	}

	if len(bytes) != 4 {
		return 0, errors.New("[]byte for uint 4 bytes must be of size 4")
	}

	return int((uint(bytes[0]) << 24) |
		(uint(bytes[1]) << 16) |
		(uint(bytes[2]) << 8) |
		(uint(bytes[3]) << 0)), nil
}

func serializeUndefined() []byte {
	return []byte{types.UNDEFINED}
}
func deserializeUndefined(buffer []byte, index int) (types.SerializableData, int, error) {
	if index+1 > len(buffer) {
		return nil, 1, errors.New("buffer too short for undefined deserialize")
	}

	if buffer[index] != types.UNDEFINED {
		return nil, 1, errors.New("wrong type for undefined")
	}

	return nil, 1, nil
}

func serializeBoolean(boolean bool) []byte {
	buffer := make([]byte, 2)
	buffer[0] = types.BOOLEAN
	if boolean {
		buffer[1] = 1
	} else {
		buffer[1] = 0
	}
	return buffer
}
func deserializeBoolean(buffer []byte, index int) (bool, int, error) {
	if index+2 > len(buffer) {
		return false, 2, errors.New("buffer too short for boolean")
	}

	if buffer[index] != types.BOOLEAN {
		return false, 2, errors.New("wrong type for boolean")
	}

	switch buffer[index+1] {
	case 1:
		return true, 2, nil
	case 0:
		return false, 2, nil
	}

	return false, 2, errors.New("wrong value for boolean")
}

func serializeString(str string) ([]byte, error) {
	strData := []byte(str)
	strSize := len(strData)
	size, err := NumberToUin4Bytes(strSize)

	if err != nil {
		return nil, errors.New("problem with str size serialize")
	}

	buffer := make([]byte, strSize+5)
	buffer[0] = types.STRING

	copy(buffer[1:5], size)
	copy(buffer[5:], strData)

	return buffer, nil
}
func deserializeString(buffer []byte, index int) (string, int, error) {
	if index+5 > len(buffer) {
		return "", 0, errors.New("buffer too short for string deserialize")
	}

	if buffer[index] != types.STRING {
		return "", 0, errors.New("wrong type for string")
	}
	index++

	size, _ := Uint4BytesToNumber(buffer[index : index+4])
	index += 4

	if index+size > len(buffer) {
		return "", size + 5, errors.New("buffer length too short to get slice of expected size")
	}

	return string(buffer[index : index+size]), size + 5, nil
}

func serializeNumber(num float64) []byte {
	buffer := make([]byte, 9)
	buffer[0] = types.NUMBER
	binary.BigEndian.PutUint64(buffer[1:], math.Float64bits(num))
	return buffer
}
func deserializeNumber(buffer []byte, index int) (float64, int, error) {
	if index+9 > len(buffer) {
		return 0, 9, errors.New("buffer too short for number deserialize")
	}

	if buffer[index] != types.NUMBER {
		return 0, 9, errors.New("wrong type for number")
	}
	index++

	bits := binary.BigEndian.Uint64(buffer[index : index+8])
	float := math.Float64frombits(bits)
	return float, 9, nil
}

func serializeBuffer(buffer []byte) ([]byte, error) {
	bufferSize := len(buffer)
	size, err := NumberToUin4Bytes(bufferSize)

	if err != nil {
		return nil, errors.New("problem with buffer size serialize")
	}

	buffer2 := make([]byte, bufferSize+5)
	buffer2[0] = types.BUFFER
	copy(buffer2[1:5], size)
	copy(buffer2[5:], buffer)
	return buffer2, nil
}
func deserializeBuffer(buffer []byte, index int) ([]byte, int, error) {
	if index+5 > len(buffer) {
		return nil, 0, errors.New("buffer too short for buffer deserialize")
	}

	if buffer[index] != types.BUFFER {
		return nil, 0, errors.New("wrong type for buffer")
	}
	index++

	size, _ := Uint4BytesToNumber(buffer[index : index+4])
	index += 4

	if index+size > len(buffer) {
		return nil, size + 5, errors.New("buffer too short for buffer deserialize")
	}

	return buffer[index : index+size], size + 5, nil
}

func serializeObject(obj any) ([]byte, error) {
	jsonBuffer, err := json.Marshal(obj)

	if err != nil {
		return nil, errors.New("failed to marshal object")
	}

	jsonBufferSize := len(jsonBuffer)
	size, err := NumberToUin4Bytes(jsonBufferSize)

	if err != nil {
		return nil, errors.New("problem with object size serialize")
	}

	buffer := make([]byte, jsonBufferSize+5)
	buffer[0] = types.OBJECT
	copy(buffer[1:5], size)
	copy(buffer[5:], jsonBuffer)
	return buffer, nil
}

func deserializeObject(buffer []byte, index int) (types.DeserializedRawObject, int, error) {
	obj := types.DeserializedRawObject{}

	if index+5 > len(buffer) {
		return obj, 0, errors.New("buffer too short for object deserialize")
	}

	if buffer[index] != types.OBJECT {
		return obj, 0, errors.New("wrong type for object")
	}
	index++

	size, _ := Uint4BytesToNumber(buffer[index : index+4])
	index += 4

	if index+size > len(buffer) {
		return obj, size + 5, errors.New("buffer too short for object deserialize")
	}

	obj.Data = buffer[index : index+size]
	return obj, size + 5, nil
}

func Serialize(data types.SerializableData) ([]byte, error) {
	serialized := ([]byte)(nil)
	err := (error)(nil)
	switch data := data.(type) {
	case nil:
		serialized = serializeUndefined()
	case bool:
		serialized = serializeBoolean(data)
	case float64:
		serialized = serializeNumber(data)
	case string:
		serialized, err = serializeString(data)
	case []byte:
		serialized, err = serializeBuffer(data)
	default:
		serialized, err = serializeObject(data)
	}

	return serialized, err
}

func MergeBuffers(buffers [][]byte) ([]byte, error) {
	if buffers == nil {
		return nil, errors.New("cannot merge nil buffers")
	}

	size := 0
	for _, buf := range buffers {
		if buf == nil {
			return nil, errors.New("received nil buffer to merge")
		}
		size += len(buf)
	}

	buffer := make([]byte, size)
	offset := 0
	for _, buf := range buffers {
		size := len(buf)
		copy(buffer[offset:offset+size], buf)
		offset += size
	}

	return buffer, nil
}

func Deserialize(buffer []byte, index int) (types.DeserializedData, error) {
	if len(buffer) == 0 || buffer == nil {
		return types.DeserializedData{}, errors.New("buffer is nil or size 0")
	}

	dataType := buffer[index]
	data := (types.SerializableData)(nil)
	size := 0
	err := (error)(nil)
	switch dataType {
	case types.UNDEFINED:
		data, size, err = deserializeUndefined(buffer, index)
	case types.BOOLEAN:
		data, size, err = deserializeBoolean(buffer, index)
	case types.STRING:
		data, size, err = deserializeString(buffer, index)
	case types.NUMBER:
		data, size, err = deserializeNumber(buffer, index)
	case types.BUFFER:
		data, size, err = deserializeBuffer(buffer, index)
	case types.OBJECT:
		data, size, err = deserializeObject(buffer, index)
	default:
		return types.DeserializedData{}, errors.New("unknown data type for buffer")
	}

	if err != nil {
		return types.DeserializedData{}, errors.New("failed to deserialize data of type: " + strconv.Itoa(int(dataType)))
	}
	return types.DeserializedData{
		Data:           data,
		Type:           dataType,
		SizeSerialized: size,
	}, nil
}

func DeserializeAll(buffer []byte) ([]types.DeserializedData, error) {
	if buffer == nil {
		return nil, errors.New("cannot deserialize all buffer nil or size 0")
	}

	data := []types.DeserializedData{}
	index := 0
	for index < len(buffer) {
		deserialized, err := Deserialize(buffer, index)
		if (err) != nil {
			return data, err
		}
		index += deserialized.SizeSerialized
		data = append(data, deserialized)
	}
	return data, nil
}

File: test/serialization.ts
import test, { suite } from "node:test";
import assert from "node:assert";
import {
    dataTypeSwitch,
    deserialize,
    deserializeBoolean,
    deserializeBuffer,
    deserializeNumber,
    deserializeObject,
    deserializeString,
    deserializeUndefined,
    getBufferSliceFromSizeData,
    mergeUint8Arrays,
    serialize,
    numberToUint4Bytes,
    serializeBoolean,
    serializeBuffer,
    serializeNumber,
    serializeObject,
    serializeString,
    serializeUndefined,
    uint4BytesToNumber
} from "../lib/bridge/serialization.ts";
import {
    BOOLEAN,
    BUFFER,
    MAX_UINT_4_BYTES,
    NUMBER,
    OBJECT,
    STRING,
    UNDEFINED
} from "../lib/@types/index.ts";

suite("bridge serialization", () => {
    test("number / uint 4 bytes", () => {
        const testData = [
            [0, new Uint8Array([0, 0, 0, 0])],
            [1, new Uint8Array([0, 0, 0, 1])],
            [256, new Uint8Array([0, 0, 1, 0])],
            [65536, new Uint8Array([0, 1, 0, 0])],
            [16777216, new Uint8Array([1, 0, 0, 0])],
            [MAX_UINT_4_BYTES, new Uint8Array([255, 255, 255, 255])]
        ] as const;

        for (const [dec, arr] of testData) {
            assert.deepEqual(numberToUint4Bytes(dec), arr);
            assert.deepEqual(uint4BytesToNumber(arr), dec);
        }

        assert.throws(() => numberToUint4Bytes(null));
        assert.throws(() => numberToUint4Bytes(undefined));
        assert.throws(() => numberToUint4Bytes(-1));
        assert.throws(() => numberToUint4Bytes(MAX_UINT_4_BYTES + 1));

        assert.throws(() => uint4BytesToNumber(null));
        assert.throws(() => uint4BytesToNumber(undefined));
        assert.throws(() => uint4BytesToNumber(new Uint8Array([1])));
    });

    test("undefined", () => {
        const serialized = new Uint8Array([UNDEFINED]);

        assert.deepEqual(serializeUndefined(), serialized);
        assert.deepEqual(serialize(undefined), serialized);
        assert.deepEqual(deserializeUndefined(serialized.buffer), {
            data: undefined,
            size: serialized.byteLength
        });
        assert.deepEqual(deserialize(serialized.buffer), {
            data: undefined,
            size: serialized.byteLength
        });

        assert.throws(() => deserializeUndefined(new Uint8Array([1]).buffer));
    });

    test("boolean", () => {
        const bool = false;
        const serialized = new Uint8Array([BOOLEAN, 0]);

        assert.deepEqual(serializeBoolean(bool), serialized);
        assert.deepEqual(serialize(bool), serialized);
        assert.deepEqual(deserializeBoolean(serialized.buffer), {
            data: bool,
            size: serialized.byteLength
        });
        assert.deepEqual(deserialize(serialized.buffer), {
            data: bool,
            size: serialized.byteLength
        });

        assert.throws(() => serializeBoolean(null));
        assert.throws(() => serializeBoolean(undefined));
        assert.throws(() => deserializeBoolean(new Uint8Array([0]).buffer));
    });

    test("string", () => {
        const str = "test";
        const serialized = new Uint8Array([
            STRING,
            ...numberToUint4Bytes(new TextEncoder().encode(str).byteLength),
            ...new TextEncoder().encode(str)
        ]);

        assert.deepEqual(serializeString(str), serialized);
        assert.deepEqual(serialize(str), serialized);
        assert.deepEqual(deserializeString(serialized.buffer), {
            data: str,
            size: serialized.byteLength
        });
        assert.deepEqual(deserialize(serialized.buffer), {
            data: str,
            size: serialized.byteLength
        });

        assert.throws(() => serializeString(null));
        assert.throws(() => serializeString(undefined));
        assert.throws(() => deserializeString(new Uint8Array([0]).buffer));
        assert.throws(() =>
            deserializeString(new Uint8Array([STRING, 0, 0, 0, 1]).buffer)
        );
    });

    test("number", () => {
        const num = 1;
        const serialized = new Uint8Array([NUMBER, 63, 240, 0, 0, 0, 0, 0, 0]);

        assert.deepEqual(serializeNumber(num), serialized);
        assert.deepEqual(serialize(num), serialized);
        assert.deepEqual(deserializeNumber(serialized.buffer), {
            data: num,
            size: serialized.byteLength
        });
        assert.deepEqual(deserialize(serialized.buffer), {
            data: num,
            size: serialized.byteLength
        });

        assert.throws(() => serializeNumber(null));
        assert.throws(() => serializeNumber(undefined));
        assert.throws(() => deserializeNumber(new Uint8Array([0]).buffer));
        assert.throws(() =>
            deserializeNumber(new Uint8Array([NUMBER, 1]).buffer)
        );
    });

    test("buffer / uint8array", () => {
        const arr = new Uint8Array([1, 2, 3, 4]);
        const serialized = new Uint8Array([
            BUFFER,
            ...numberToUint4Bytes(arr.byteLength),
            ...arr
        ]);

        assert.deepEqual(serializeBuffer(arr), serialized);
        assert.deepEqual(serialize(arr), serialized);
        assert.deepEqual(deserializeBuffer(serialized.buffer), {
            data: arr,
            size: serialized.byteLength
        });
        assert.deepEqual(deserialize(serialized.buffer), {
            data: arr,
            size: serialized.byteLength
        });

        assert.throws(() => deserializeBuffer(new Uint8Array([0]).buffer));
        assert.throws(() =>
            deserializeBuffer(new Uint8Array([0, 0, 0, 1, 0]).buffer)
        );
        assert.throws(() =>
            deserializeBuffer(new Uint8Array([BUFFER, 0, 0, 0, 1]).buffer)
        );
    });

    test("object", () => {
        const obj = {
            test: "foo"
        };
        const serialized = new Uint8Array([
            OBJECT,
            ...numberToUint4Bytes(
                new TextEncoder().encode(JSON.stringify(obj)).byteLength
            ),
            ...new TextEncoder().encode(JSON.stringify(obj))
        ]);

        assert.deepEqual(serializeObject(obj), serialized);
        assert.deepEqual(serialize(obj), serialized);
        assert.deepEqual(deserializeObject(serialized.buffer), {
            data: obj,
            size: serialized.byteLength
        });
        assert.deepEqual(deserialize(serialized.buffer), {
            data: obj,
            size: serialized.byteLength
        });

        assert.throws(() => serializeObject(null));
        assert.throws(() => serializeObject(undefined));
        assert.throws(() => deserializeObject(new Uint8Array([0]).buffer));
        assert.throws(() =>
            deserializeObject(new Uint8Array([OBJECT, 0, 0, 0, 2, 0, 0]).buffer)
        );
    });

    test("object (array)", () => {
        const arr = [1, 2, 3, 4];
        const serialized = new Uint8Array([
            OBJECT,
            ...numberToUint4Bytes(
                new TextEncoder().encode(JSON.stringify(arr)).byteLength
            ),
            ...new TextEncoder().encode(JSON.stringify(arr))
        ]);

        assert.deepEqual(serializeObject(arr), serialized);
        assert.deepEqual(serialize(arr), serialized);
        assert.deepEqual(deserializeObject(serialized.buffer), {
            data: arr,
            size: serialized.byteLength
        });
        assert.deepEqual(deserialize(serialized.buffer), {
            data: arr,
            size: serialized.byteLength
        });
    });

    test("data type switch", () => {
        assert.deepEqual(dataTypeSwitch(undefined), UNDEFINED);
        assert.deepEqual(dataTypeSwitch(false), BOOLEAN);
        assert.deepEqual(dataTypeSwitch(""), STRING);
        assert.deepEqual(dataTypeSwitch(0), NUMBER);
        assert.deepEqual(dataTypeSwitch(new Uint8Array()), BUFFER);
        assert.deepEqual(dataTypeSwitch({}), OBJECT);
        assert.deepEqual(dataTypeSwitch([]), OBJECT);
        assert.deepEqual(dataTypeSwitch(new ArrayBuffer()), null);
    });

    test("merge uint8arrays", () => {
        assert.deepEqual(mergeUint8Arrays(), new Uint8Array());
        assert.deepEqual(
            mergeUint8Arrays(new Uint8Array([0])),
            new Uint8Array([0])
        );
        assert.deepEqual(
            mergeUint8Arrays(new Uint8Array([0]), new Uint8Array([0])),
            new Uint8Array([0, 0])
        );
        assert.throws(() => mergeUint8Arrays(null));
    });

    test("get slice of size from buffer", () => {
        assert.deepEqual(
            getBufferSliceFromSizeData(new Uint8Array([0, 0, 0, 1, 1]).buffer),
            {
                slice: new Uint8Array([1]).buffer,
                size: 5
            }
        );

        assert.deepEqual(
            getBufferSliceFromSizeData(new Uint8Array([0, 0, 0, 0, 1]).buffer),
            {
                slice: new Uint8Array().buffer,
                size: 4
            }
        );

        assert.throws(() =>
            getBufferSliceFromSizeData(new Uint8Array([0, 0, 0, 2, 1]).buffer)
        );
    });
});
